contracts/utils/math/Math.sol

```diff
@@ -167,35 +167,7 @@ library Math {
         // This gives `2**k < a <= 2**(k+1)` → `2**(k/2) <= sqrt(a) < 2 ** (k/2+1)`.
         // Using an algorithm similar to the msb computation, we are able to compute `result = 2**(k/2)` which is a
         // good first approximation of `sqrt(a)` with at least 1 correct bit.
-        uint256 result = 1;
-        uint256 x = a;
-        if (x >> 128 > 0) {
-            x >>= 128;
-            result <<= 64;
-        }
-        if (x >> 64 > 0) {
-            x >>= 64;
-            result <<= 32;
-        }
-        if (x >> 32 > 0) {
-            x >>= 32;
-            result <<= 16;
-        }
-        if (x >> 16 > 0) {
-            x >>= 16;
-            result <<= 8;
-        }
-        if (x >> 8 > 0) {
-            x >>= 8;
-            result <<= 4;
-        }
-        if (x >> 4 > 0) {
-            x >>= 4;
-            result <<= 2;
-        }
-        if (x >> 2 > 0) {
-            result <<= 1;
-        }
+        uint256 result = 1 << (log2(a) / 2);
```

Comment:
We can replace `1 << (log2(a) / 2)` by `1 << (log2(a) >> 2)`.

Reply:
That is interresting, but the correct formula would be `1 << (log2(a) >> 1)`

Reply:
Ohh, sorry. It >>1.

Reply:
Before change that I would check if Solidity does not make that optimization on its own.

Reply:
tested both. using `>> 1` saves 54 gas compared to ```/ 2```
(0.8.16 with optimizations)

Reply:
@Amxx Because It happen due to solidity checking underflow that. But When we are use shift operator it bypass that check.

Reply:
Overflow check make sens for +, - and * ... not for /
The only reasonable check is that we don't divide by 0 ... but there is no reason to check that here.